[WIP] File formats

hamishwillee opened the pull request:
This adds file format doc from https://dev.qgroundcontrol.com/en/file_formats/ - updated.
Some of these things were already documented (partially) so I am assuming we want these to be used as a standard.

@DonLakeFlyer @LorenzMeier Once stuff is standardised here, would link to it from QGC Devguide, removing duplication.

Some questions:
1. Is the survey format intended to also be "standardized"?
2. I saved a mission plan out of QGC where I had also defined a geofence and rally points. The geofence and rally point lists in the saved plan were empty lists. Do I need to do anything else to export these?
3. What is the geofence/rally point "intention"? Are these associated with the mission (- ie if we get RTL vehicle goes to THIS rally point rather than some other set) or are they just part of a single set of points used by the vehicle for everything?
4. Similarly, it looks like you can generate standalone geofence and rally point plans. How do I do this? If we don't allow generation, do we support import of them?

I will add few other questions inline.

hamishwillee commented on en/SUMMARY.md:
```diff
@@ -21,7 +21,11 @@
+    * [Mission Plan](file_formats/mission_plan.md)
+    * [GeoFence Plan](file_formats/geofence.md)
+    * [Rally/Safe Points Plan](file_formats/rally_points.md)
+    * [Mission Text Plan (Deprecated)](file_formats/mission_plan_plain_text.md)
```
Assume I need survey format too?

DonLakeFlyer replied:
Both Survey and Corridor Scan.

hamishwillee commented on en/file_formats/geofence.md:
```diff
@@ -0,0 +1,80 @@
+    "groundStation": "QGroundControl",
+    "parameters": [
+        {
+            "compId": 1,
```
What is compID?

DonLakeFlyer replied:
Component ID. Are you running an old build, I thought I had pulled out the parameters sections. It is no longer used.

hamishwillee commented on en/file_formats/geofence.md:
```diff
@@ -0,0 +1,80 @@
+    "parameters": [
+        {
+            "compId": 1,
+            "name": "FENCE_ENABLE",
```
@DonLakeFlyer All these params are for PX4 geofence - ie cylindrical geofence defined using parameters.  What happens with ArduPilot and others?
Or is this PX4 specific (in which case, presumably not part of the standard?)

DonLakeFlyer replied:
It looks like you are running an old build?

hamishwillee commented on en/file_formats/geofence.md:
```diff
@@ -0,0 +1,80 @@
+            "value": 0
+        }
+    ],
+    "polygon": [
```
Is there a way to specify the altitude of the polygon?

hamishwillee commented on en/file_formats/geofence.md:
```diff
@@ -0,0 +1,80 @@
+    ],
+    "polygon": [
+        [
+            47.634457973002796,
```
Are these ordered? Or are they arranged automatically such that there is no cross-over of the borders?

hamishwillee commented on en/file_formats/geofence.md:
```diff
@@ -0,0 +1,80 @@
+            -122.08958864075316
+        ],
+        [
+            47.634371216366716,
```
Do these just map to MISSION_ITEM x, y values?

hamishwillee commented on en/file_formats/rally_points.md:
```diff
@@ -0,0 +1,51 @@
+            50
+        ],
+        [
+            47.634244700000004,
```
Do points just map to https://mavlink.io/en/messages/common.html#MISSION_ITEM  x, y, z values?

hamishwillee commented on en/file_formats/mission_plan.md:
```diff
@@ -0,0 +1,168 @@
+cruiseSpeed | The default cruise speed for the mission (metres/second).
+hoverSpeed | The default hover speed for the mission (metres/second). <!-- what is hover speed? Only for Planes? What if on a still-hover system like MC) -->
+plannedHomePosition | The planned home position to show on the map when you are editing the mission. Values with array are latitude, longitude and altitude.
+items | The list of mission item objects associated with the mission. Each item will either be a [SimpleItem](#SimpleItem) or [ComplexItem](#ComplexItem). <!-- can you mix them? -->
```
Can the list of items mix complex and simple items in the one plan?

DonLakeFlyer replied:
Cruise speed is fixed wing or VTOL forward flight. Hover speed is multi-rotor speed.

> The list of mission item objects associated with the mission. Each item will either be a [SimpleItem](#SimpleItem) or [ComplexItem](#ComplexItem).

> Can the list of items mix complex and simple items in the one plan?

Yes.

hamishwillee commented on en/file_formats/mission_plan.md:
```diff
@@ -0,0 +1,168 @@
+
+## GeoFence Object {#geoFence}
+
+The (optional) geofence information associated with this flight plan has the same format as defined in [GeoFence Plan](../file_formats/geofence.md)
```
Is this and Rally point doc correct - ie you could copy-paste in the geofence from the other doc, remove the file specific stuff and it would just "work"?

Note: My test of export from QGC left this list empty.

DonLakeFlyer replied:
The GeoFence and Rally Point file formats docs are both completely out of date.

DonLakeFlyer commented:
The QGC file formats shouldn't be part of the mavlink spec since they are ground station specific. They only relate to QGC.

DonLakeFlyer commented:
> I saved a mission plan out of QGC where I had also defined a geofence and rally points. The geofence and rally point lists in the saved plan were empty lists. Do I need to do anything else to export these?

You don't need to do anything. If there were not saved it's a bug.


> What is the geofence/rally point "intention"? Are these associated with the mission (- ie if we get RTL vehicle goes to THIS rally point rather than some other set) or are they just part of a single set of points used by the vehicle for everything?

They are associated with the Plan. A "Plan" is a mission, geofence and rally point combination.

> Similarly, it looks like you can generate standalone geofence and rally point plans. How do I do this? If we don't allow generation, do we support import of them?

You should be able to do this. You just don't add any mission items.

hamishwillee commented:
@LorenzMeier Our current [mission protocol docs](https://mavlink.io/en/protocol/mission.html#mission-file-format) includes info about the plan file format. I created this PR to update the information based on more detailed documentation in the QGC Devguide.

BUT, according to @DonLakeFlyer  (Thanks Don!)

> The QGC file formats shouldn't be part of the mavlink spec since they are ground station specific. They only relate to QGC.

Is the plan to make these part of the standard, and encourage other GCS to adopt them? If so, we probably need to work with Don to get agreement because otherwise he is going to feel he can break that standard as needed. Do we need any formal buy in from the wider team?

Note, I think this is a good idea, if only because no one else has a standard, and being able to store and manipulate missions offline is useful for both GCS and developer APIs.

DonLakeFlyer commented:
> If so, we probably need to work with Don to get agreement because otherwise he is going to feel he can break that standard as needed.

QGC can't "break" the QGC standard. If it does it will no longer be able to load older files. If the file formats change, QGC is always changed in a backward compatible way such that it can load old and new formats. I update the QGC file formats as needed to implement new features.

It's fine for people to create the .plan files manually by hand following the documented QGC formats. My point is that this has nothing to do with mavlink since that is a wire protocol. It only relates to QGroundControl.

hamishwillee commented:
@DonLakeFlyer The problem is that the MAVLink docs *already* include a version of the plan file. While I understand your point, I need to understand the intention.

I'd also point out that MAVLink is whatever the project says it is. You could already argue that it is more than just a wire protocol, since it includes generators and a reference ground station - QGC.

hamishwillee commented:
FYI, while continuing this discussion, I have moved to daily build and updated QGC docs (in progress): https://github.com/mavlink/qgc-dev-guide/pull/43

hamishwillee commented:
FYI, Dev Call meeting agreed that defining file formats as part of MAVLink project is *desirable*, and should be discussed over coming months. However this PR is unlikely to be the final agree format. 

I am therefore closing.
